auth/api/iam/openid4vp.go

```diff
@@ -2,75 +2,264 @@ package iam
 
 import (
 	"bytes"
+	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"github.com/labstack/echo/v4"
-	"github.com/nuts-foundation/nuts-node/core"
-	"html/template"
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/jsonld"
+	"github.com/nuts-foundation/nuts-node/vcr"
+	"github.com/nuts-foundation/nuts-node/vcr/credential"
+	"github.com/nuts-foundation/nuts-node/vcr/holder"
 	"net/http"
+	"net/url"
+	"strings"
 )
 
-// openID4VP implements verifiable presentation exchanges as specified by https://openid.net/specs/openid-4-verifiable-presentations-1_0.html.
-type openID4VP struct {
-	sessions      *SessionManager
-	authzTemplate *template.Template
-}
+const clientIDParam = "client_id"
+const responseTypeParam = "response_type"
+const scopeParam = "scope"
+const stateParam = "state"
+const redirectURIParam = "redirect_uri"
+const presentationDefParam = "presentation_definition"
+const presentationDefUriParam = "presentation_definition_uri"
+const presentationSubmissionParam = "presentation_submission"
+const clientMetadataParam = "client_metadata"
+const clientMetadataURIParam = "client_metadata_uri"
+const clientIDSchemeParam = "client_id_scheme"
+const responseModeParam = "response_mode"
+const vpTokenParam = "vp_token"
 
-func (a openID4VP) Routes(router core.EchoRouter) {
-	router.Add(http.MethodPost, "/public/oauth2/:did/authz_consent", a.handleAuthConsent)
+// createPresentationRequest creates a new Authorization Request as specified by OpenID4VP: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html.
+// It is sent by a verifier to a wallet, to request one or more verifiable credentials as verifiable presentation from the wallet.
+func (r *Wrapper) sendPresentationRequest(ctx context.Context, response http.ResponseWriter, scope string,
+	redirectURL url.URL, verifierIdentifier url.URL, walletIdentifier url.URL) error {
+	// TODO: Lookup wallet metadata for correct authorization endpoint. But for Nuts nodes, we derive it from the walletIdentifier
+	authzEndpoint := walletIdentifier.JoinPath("/authorize")
+	params := make(map[string]string)
```

Comment:
can probably be simplified by marshaling a struct? TODO for now I think

Reply:
Later on, yes, when we support form post as well. Now they're URL query parameters, so a map is more practical to iterate over.